provider/machinepool/machine_pool_datasource.go

```diff
@@ -0,0 +1,199 @@
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"key": schema.StringAttribute{
+							Description: "Taints key",
+							Computed:    true,
+						},
+						"value": schema.StringAttribute{
+							Description: "Taints value",
+							Computed:    true,
+						},
+						"schedule_type": schema.StringAttribute{
+							Description: "Taints schedule type",
+							Computed:    true,
+						},
+					},
+				},
+				Computed: true,
+			},
+			"labels": schema.MapAttribute{
+				Description: "Labels for the machine pool. Format should be a comma-separated list of 'key = value'." +
+					" This list will overwrite any modifications made to node labels on an ongoing basis.",
+				ElementType: types.StringType,
+				Computed:    true,
+			},
+			"multi_availability_zone": schema.BoolAttribute{
+				Description: "Create a multi-AZ machine pool for a multi-AZ cluster (default is `true`). " + common.ValueCannotBeChangedStringDescription,
+				Computed:    true,
+			},
+			"availability_zone": schema.StringAttribute{
+				Description: "Select the availability zone in which to create a single AZ machine pool for a multi-AZ cluster. " + common.ValueCannotBeChangedStringDescription,
+				Computed:    true,
+			},
+			"availability_zones": schema.ListAttribute{
+				Description: "Availability zones. ",
+				ElementType: types.StringType,
+				Computed:    true,
+			},
+			"subnet_id": schema.StringAttribute{
+				Description: "Select the subnet in which to create a single AZ machine pool for BYO-VPC cluster. " + common.ValueCannotBeChangedStringDescription,
+				Computed:    true,
+			},
+			"aws_subnet_ids": schema.ListAttribute{
+				Description: "AWS subnet IDs. ",
+				ElementType: types.StringType,
+				Computed:    true,
+			},
+			"disk_size": schema.Int64Attribute{
+				Description: "Root disk size, in GiB. " + common.ValueCannotBeChangedStringDescription,
+				Computed:    true,
+			},
+			"aws_additional_security_group_ids": schema.ListAttribute{
+				Description: "AWS additional security group ids. " + common.ValueCannotBeChangedStringDescription,
+				ElementType: types.StringType,
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func (r *MachinePoolDatasource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	// Get the current state:
+	state := &MachinePoolState{}
+	diags := req.Config.Get(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	notFound, diags := readState(ctx, state, r.collection)
+	if notFound {
+		// If we can't find the machine pool, it was deleted. Remove if from the
+		// state and don't return an error so the TF apply() will automatically
+		// recreate it.
+		tflog.Warn(ctx, fmt.Sprintf("machine pool (%s) of cluster (%s) not found, removing from state",
+			state.ID.ValueString(), state.Cluster.ValueString(),
+		))
+		resp.State.RemoveResource(ctx)
+		return
+	}
```

Comment:
This shouldn't have the same behavior like in the resource
You are removing here resource while you actually handle data source
I think it just should return error the the data source doesn't exist, but not sure
What do you think?